components/region_cache_memory_engine/src/write_batch.rs

```diff
@@ -28,13 +60,88 @@ impl RegionCacheWriteBatch {
         self.sequence_number = Some(seq);
         Ok(())
     }
+
+    fn write_impl(&mut self, seq: u64) -> Result<()> {
+        self.buffer
+            .iter()
+            .map(|e| (e.cf.as_str(), e.encode(seq)))
+            .try_for_each(|(cf, (key, value))| (self.apply_cb)(cf, key, value))
+    }
 }
 
+#[derive(Clone, Debug)]
+enum CacheWriteBatchEntryMutation {
+    PutValue(Bytes),
+    Deletion,
+}
+
+impl CacheWriteBatchEntryMutation {
+    fn encode(&self, key: &[u8], seq: u64) -> (Bytes, Bytes) {
+        match self {
+            CacheWriteBatchEntryMutation::PutValue(value) => {
+                (encode_key(key, seq, ValueType::Value), value.clone())
+            }
+            CacheWriteBatchEntryMutation::Deletion => {
+                (encode_key(key, seq, ValueType::Deletion), Bytes::new())
+            }
+        }
+    }
+    fn data_size(&self) -> usize {
+        match self {
+            CacheWriteBatchEntryMutation::PutValue(value) => value.len(),
+            CacheWriteBatchEntryMutation::Deletion => 0,
+        }
+    }
+}
 #[derive(Clone, Debug)]
 struct RegionCacheWriteBatchEntry {
     cf: String,
```

Comment:
i should change this to be a usize and call cf_to_id  in put_cf/delete_cf() to avoid partial failures.